// vuepress-theme-reco/lib/client/composables/useSeriesItems.js
import { inject } from 'vue';
import { useRoute } from 'vue-router';
import { isArray, isPlainObject, isString, resolveLocalePath, } from '@vuepress/shared';
import { useNavLink } from './useNavLink';
export const seriesItemsSymbol = Symbol('seriesItems');
export const useSeriesItems = () => {
    const seriesItems = inject(seriesItemsSymbol);
    if (!seriesItems) {
        throw new Error('useSeriesItems() is called without provider.');
    }
    return seriesItems;
};
export const resolveSeriesItems = (frontmatter, themeLocal) => {
    var _a, _b;
    // get series config from frontmatter > themeConfig
    const seriesConfig = (_b = (_a = frontmatter.series) !== null && _a !== void 0 ? _a : themeLocal.series) !== null && _b !== void 0 ? _b : 'auto';
    // resolve series items according to the config
    if (frontmatter.home || seriesConfig === false) {
        return [];
    }
    if (isArray(seriesConfig)) {
        return resolveArraySeriesItems(seriesConfig);
    }
    if (isPlainObject(seriesConfig)) {
        return resolveMultiSeriesItems(seriesConfig);
    }
    return [];
};
/**
 * Resolve series items if the config is an array
 */
export const resolveArraySeriesItems = (seriesConfig) => {
    const handleChildItem = (item) => {
        let childItem;
        if (isString(item)) {
            childItem = useNavLink(item);
        }
        else {
            childItem = item;
        }
        if (childItem.isGroup && childItem.children) {
            return {
                ...childItem,
                children: childItem.children.map(handleChildItem),
            };
        }
        return childItem;
    };
    return seriesConfig.map((item) => {
        if (isString(item)) {
            return useNavLink(item);
        }
        return {
            ...item,
            children: item.children.map(handleChildItem),
        };
    });
};
/**
 * Resolve series items if the config is a key -> value (path-prefix -> array) object
 */
export const resolveMultiSeriesItems = (seriesConfig) => {
    var _a;
    const route = useRoute();
    const seriesPath = resolveLocalePath(seriesConfig, route.path);
    const matchedSeriesConfig = (_a = seriesConfig[seriesPath]) !== null && _a !== void 0 ? _a : [];
    return resolveArraySeriesItems(matchedSeriesConfig);
};
